{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Jv3XFoGKCzXB"
   },
   "source": [
    "# 1. Introducción a STRING."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "4gXqegHEC1xZ"
   },
   "source": [
    "## 1.1. ¿Qué es?\n",
    "STRING es una base de datos que recopila información sobre las relaciones entre proteínas. Esta herramienta ofrece un grafo que establece las compatibilidades, así como similitudes entre las moléculas. Para ello, STRING hace uso de una amplia fuente de datos, así como mecanismos de análisis, para ofrecer detalles sobre estos elementos."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jfGQECe9C4vH"
   },
   "source": [
    "\n",
    "## 1.2. ¿Cómo funciona?\n",
    "STRING proporciona detalles sobre las relaciones entre proteínas a través de cuatro fuentes de datos principales, enriquecidas con mecanismos adicionales y tecnologías de análisis y predicción. Las fuentes de datos que alimentan la red de proteínas, así como algunos recursos adicionales, son:\n",
    "- Datos sobre las fusiones génicas.\n",
    "- Información del contexto genómico de las proteínas.\n",
    "- Vecindad entre proteínas.\n",
    "- Los perfiles filogénicos dentro del árbol de la vida.\n",
    "- Similitudes entre expresiones genéticas.\n",
    "- Experimentos sobre interacciones entre proteínas.\n",
    "- Datasets curados.\n",
    "- Pathway moleculares.\n",
    "- Mapas bioquímicos.\n",
    "- Machine Learning.\n",
    "- Deep Learning.\n",
    "\n",
    "Todo el conocimiento extraído y almacenado en la base de datos de STRING es accesible a través del portal web de la plataforma. Del mismo modo, con el objetivo de facilitar la integración de los resultados en implementaciones propias, STRING proporciona una API que permite emprender consultas de manera programática. Como resultado, en estos ejercicios se repasaran las operaciones, así como las interacciones, más relevantes del servicio."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "zEai5f6-C7ea"
   },
   "source": [
    "## 1.3. ¿Cómo resolver este notebook?\n",
    "Este _notebook_ propone un recorrido breve pero profundo de las posibilidades de la API REST de STRING. Para ello, los ejercicios proponen dos enfocques principales: emprender solicitudes a la API REST a través de la URL y realizar consultas mediante la librería stringdb."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "6pvRPjQ4C9rH"
   },
   "source": [
    "# 2. Acceso programático a los recursos de STRING.\n",
    "STRING proporciona una API accesible mediante peticiones HTTP, así como una librería de Python para interacturar con los _endpoints_. A continuación, se propone un conjunto de ejercicios basados en la invocación directa de la API, así como en el uso de la librería."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "47_QtzC3DA0k"
   },
   "source": [
    "## 2.1. Consultas mediante URL.\n",
    "La API Rest de STRING propone un conjunto de _endpoints_ para interactuar. Estos puntos de consulta permiten indicar el tipo de información que se quiere obtener. De esta manera, se distingue un total de nueve _endpoints_. Puedes encontrar todos los detalles en https://string-db.org/cgi/help.pl?subpage=api.\n",
    "\n",
    "---\n",
    "\n",
    "https://string-db.org/api\n",
    "  /FORMAT\n",
    "  /METHODS\n",
    "\n",
    "**/FORMAT**\n",
    "\n",
    "Las opciones posibles para el formato de salida de la petición dependen activamente del _endpoint_ que se consulte. A continuación, junto con cada método, se detallará el conjunto de opciones válidas.\n",
    "\n",
    "**/METHODS**\n",
    "\n",
    "> get_string_ids\n",
    "\n",
    "Formatos: tsv, tsv-no-header, json, xml.\n",
    "\n",
    "Parámetros:\n",
    "- identifiers= Indica el identificador o identificadores que se involucrarán en la consulta, separados por %0d. Ejemplo: identifiers=DRD1_HUMAN%0dDRD2_HUMAN.\n",
    "\n",
    "- echo_query= [Opcional] Añade una columna con los datos proporcionadas como parámetro. Se indica mediante 0 o 1.\n",
    "- limit=\t[Opcional] Limita el número de coincidencias por identificador proporcionado. Las mejores coincidencias figurarán primero.\n",
    "- species=\t[Opcional] Indica el identificador taxon-NCBI ID de especies. Por ejemplo, en el caso del homo sapiens es 9606. Podrás encontrar más datas en https://string-db.org/cgi/input.pl?input_page_active_form=organisms.\n",
    "- caller_identity= [Opcional]\tIdentificación de la petición.\n",
    "\n",
    "Como salida de la petición, se recibirán los siguientes campos de información:\n",
    "- queryItem [Opcional]. ID del elemento solicitado.\n",
    "- queryIndex. Posición de la proteína en la consulta. Comienza en cero.\n",
    "- strinId. Identificador en STRING.\n",
    "- ncbiTaxonId. Identificador en taxon-NCBI.\n",
    "- taxonName. Nombre de la especie.\n",
    "- preferredName. Nombre común de la proteína.\n",
    "- annotation. Anotaciones sobre la proteína.\n",
    "\n",
    "> network?PARAMETERS\n",
    "\n",
    "Formatos: image, highres_image, avg\n",
    "\n",
    "Parámetros:\n",
    "- identifiers= Indica el identificador o identificadores que se involucrarán en la consulta, separados por %0d. Ejemplo: identifiers=DRD1_HUMAN%0dDRD2_HUMAN.\n",
    "- species= [Opcional] Indica el identificador taxon-NCBI ID de especies. Por ejemplo, en el caso del homo sapiens es 9606. Podrás encontrar más datas en https://string-db.org/cgi/input.pl?input_page_active_form=organisms.\n",
    "- add_color_nodes= [Opcional] Establece una gama de colores en los nodos basado en la puntuación de las proteínas indicadas como argumento.\n",
    "- add_white_nodes= [Opcional] Establece nodos blancos basados en la puntuación de las proteínas indicadas como argumento.\n",
    "- required_score= [Opcional] Umbral de relevancia de las relaciones entre las proteínas a incluir. De esta manera, el valor se gradúa entre 0 y 1000.\n",
    "- network_flavor= [Opcional] Estilo de las aristas de la red: evidence, confidence (default), actions.\n",
    "- network_type= [Opcional] Tipo de relación que se quiere incluir: functional, physical.\n",
    "- hide_node_labels= [Opcional] Indicando 0 o 1, permite configurar si se muestra o no el nombre de las proteínas en la red.\n",
    "- hide_disconnected_nodes= [Opcional] 0/1.\n",
    "- show_query_node_labels= [Opcional] 0/1. Permite indicar si la etiqueta de las proteínas indicadas como argumento debe mostrarse.\n",
    "- block_structure_pics_in_bubbles [Opcional] 0/1.\n",
    "- caller_identity= [Opcional]\tIdentificación de la petición.\n",
    "\n",
    "Como salida de la petición, se recibirán los siguientes campos de información:\n",
    "\n",
    "- imagen.\n",
    "\n",
    "> get_link?PARAMETERS\n",
    "\n",
    "Proporciona un enlace en STRING para la imagen de la red solicitada.\n",
    "\n",
    "Formato: tsv, tsv-no-header, json, xml.\n",
    "\n",
    "Parámetros:\n",
    "- identifiers= Indica el identificador o identificadores que se involucrarán en la consulta, separados por %0d. Ejemplo: identifiers=DRD1_HUMAN%0dDRD2_HUMAN.\n",
    "- species= [Opcional] Indica el identificador taxon-NCBI ID de especies. Por ejemplo, en el caso del homo sapiens es 9606. Podrás encontrar más datas en https://string-db.org/cgi/input.pl?input_page_active_form=organisms.\n",
    "- add_color_nodes= [Opcional] Establece una gama de colores en los nodos basado en la puntuación de las proteínas indicadas como argumento.\n",
    "- add_white_nodes= [Opcional] Establece nodos blancos basados en la puntuación de las proteínas indicadas como argumento.\n",
    "- required_score= [Opcional] Umbral de relevancia de las relaciones entre las proteínas a incluir. De esta manera, el valor se gradúa entre 0 y 1000.\n",
    "- network_flavor= [Opcional] Estilo de las aristas de la red: evidence, confidence (default), actions.\n",
    "- network_type= [Opcional] Tipo de relación que se quiere incluir: functional, physical.\n",
    "- hide_node_labels= [Opcional] Indicando 0 o 1, permite configurar si se muestra o no el nombre de las proteínas en la red.\n",
    "- hide_disconnected_nodes= [Opcional] 0/1.\n",
    "- show_query_node_labels= [Opcional] 0/1. Permite indicar si la etiqueta de las proteínas indicadas como argumento debe mostrarse.\n",
    "- block_structure_pics_in_bubbles [Opcional] 0/1.\n",
    "- caller_identity= [Opcional]\tIdentificación de la petición.\n",
    "\n",
    "Como salida de la petición, se recibirán los siguientes campos de información:\n",
    "- url. Link estable al rescurso.\n",
    "\n",
    "> network?PARAMETERS\n",
    "\n",
    "Devuelve los detalles de las relaciones de las proteínas proporcionadas.\n",
    "\n",
    "Formato: tsv, tsv-no-header, json, xml.\n",
    "\n",
    "Parámetros:\n",
    "- identifiers= Indica el identificador o identificadores que se involucrarán en la consulta, separados por %0d. Ejemplo: identifiers=DRD1_HUMAN%0dDRD2_HUMAN.\n",
    "- species= [Opcional] Indica el identificador taxon-NCBI ID de especies. Por ejemplo, en el caso del homo sapiens es 9606. Podrás encontrar más datas en https://string-db.org/cgi/input.pl?input_page_active_form=organisms.\n",
    "- required_score= [Opcional] Umbral de relevancia de las relaciones entre las proteínas a incluir. De esta manera, el valor se gradúa entre 0 y 1000.\n",
    "- network_type= [Opcional] Tipo de relación que se quiere incluir: functional, physical.\n",
    "- add_nodes= [Opcional] Se añade un número adicional de proteínas a la red relacionadas con el parámetro de entrada. Cuando se indica únicamente una sola proteína, este valor se establece en 10. Por otro lado, si se indica una colección de proteínas, este valor permanecerá a cero por defecto.\n",
    "- show_query_node_labels= [Opcional] 0/1. Permite indicar si la etiqueta de las proteínas indicadas como argumento debe mostrarse.\n",
    "- caller_identity= [Opcional]\tIdentificación de la petición.\n",
    "\n",
    "Como salida de la petición, se recibirán los siguientes campos de información:\n",
    "- stringId_A.\n",
    "- stringId_B.\n",
    "- preferredName_A.\n",
    "- preferredName_B.\n",
    "- ncbiTaxonId.\n",
    "- score. Puntuación combinada de la red.\n",
    "- nscore. Puntuación de vecindad genética de la red.\n",
    "- fscore. Puntuación de fusión genética.\n",
    "- pscore. Puntuación del perfil filogenético de la red.\n",
    "- ascore. Puntuación de coexpresión de la red.\n",
    "- escore. Puntuación de experimentación.\n",
    "- dscore. Puntuación de base de datos.\n",
    "- tscore. Text mining score.\n",
    "\n",
    "> homology?PARAMETERS\n",
    "\n",
    "Devuelve proteínas similares a las proporcionadas.\n",
    "\n",
    "Formato: tsv, tsv-no-header, json, xml.\n",
    "\n",
    "Parámetros:\n",
    "- identifiers= Indica el identificador o identificadores que se involucrarán en la consulta, separados por %0d. Ejemplo: identifiers=DRD1_HUMAN%0dDRD2_HUMAN.\n",
    "- species= [Opcional] Indica el identificador taxon-NCBI ID de especies. Por ejemplo, en el caso del homo sapiens es 9606. Podrás encontrar más datas en https://string-db.org/cgi/input.pl?input_page_active_form=organisms.\n",
    "- caller_identity= [Opcional]\tIdentificación de la petición.\n",
    "\n",
    "Como salida de la petición, se recibirán los siguientes campos de información:\n",
    "\n",
    "- ncbiTaxonId_A.\n",
    "- ncbiTaxonId_B.\n",
    "- stringId_A.\n",
    "- stringId_B.\n",
    "- bitscore. Smith-Waterman alignment bit score.\n",
    "\n",
    "Puedes encontrar todos los detalles del resto de _endpoints_ en https://string-db.org/cgi/help.pl?subpage=api."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import shutil\n",
    "import requests\n",
    "import pandas as pd\n",
    "import stringdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "G3XHSKnJnVcB"
   },
   "source": [
    "### 2.1.1. Obtén el nombre común de proteínas a partir de su identificador.\n",
    "ID de proteínas: 'Smo'.\n",
    "\n",
    "Especie: Mus Musculus.\n",
    "\n",
    "¿Cómo podemos hacer la búsqueda según su nombre?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "n6SFIiCHnfXh"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/get_string_ids?identifiers=Smo&species=10090\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>queryIndex</th>\n",
       "      <th>queryItem</th>\n",
       "      <th>stringId</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>taxonName</th>\n",
       "      <th>preferredName</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Smo</td>\n",
       "      <td>10090.ENSMUSP00000001812</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>Smo</td>\n",
       "      <td>Smoothened homolog; G protein-coupled receptor...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   queryIndex queryItem                  stringId  ncbiTaxonId     taxonName  \\\n",
       "0           0       Smo  10090.ENSMUSP00000001812        10090  Mus musculus   \n",
       "\n",
       "  preferredName                                         annotation  \n",
       "0           Smo  Smoothened homolog; G protein-coupled receptor...  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_url='https://string-db.org/api/json/'\n",
    "\n",
    "url_request = base_url + \"get_string_ids?identifiers=Smo&species=10090\"\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request).json()\n",
    "df = pd.DataFrame(response)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "id": "sbsMvVrT-NZV"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/get_string_ids?identifiers=10090\n"
     ]
    }
   ],
   "source": [
    "# ¿Cómo podemos hacer la búsqueda según su identificador?\n",
    "url_request = base_url + \"get_string_ids?identifiers=10090\"\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "# .."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "YYSbuVZ_niUT"
   },
   "source": [
    "### 2.1.2. ¿A qué especies pertenecen las siguientes proteínas?.\n",
    "ID de proteínas: 1, 2, 3, 4, 5. Recuerda utilizar %0d para separar los valores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "BDnBwqwHnno5"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/get_string_ids?identifiers=1%0d2%0d3%0d4%0d5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>queryIndex</th>\n",
       "      <th>queryItem</th>\n",
       "      <th>stringId</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>taxonName</th>\n",
       "      <th>preferredName</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>7227.FBpp0081985</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CG3281</td>\n",
       "      <td>LD47774p; Zinc ion binding; DNA-binding transc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>7227.FBpp0074558</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>AP-1-2beta</td>\n",
       "      <td>Adaptor Protein complex 1/2, beta subunit (AP-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>7227.FBpp0112135</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>CG12213</td>\n",
       "      <td>Uncharacterized protein, isoform C.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>7227.FBpp0077119</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Traf4</td>\n",
       "      <td>TNF-receptor-associated factor 4 (Traf4) encod...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>7227.FBpp0072162</td>\n",
       "      <td>7227</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>Zfrp8</td>\n",
       "      <td>Zinc finger protein RP-8 (Zfrp8) encodes a hig...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   queryIndex queryItem          stringId  ncbiTaxonId  \\\n",
       "0           0         1  7227.FBpp0081985         7227   \n",
       "1           1         2  7227.FBpp0074558         7227   \n",
       "2           2         3  7227.FBpp0112135         7227   \n",
       "3           3         4  7227.FBpp0077119         7227   \n",
       "4           4         5  7227.FBpp0072162         7227   \n",
       "\n",
       "                 taxonName preferredName  \\\n",
       "0  Drosophila melanogaster        CG3281   \n",
       "1  Drosophila melanogaster    AP-1-2beta   \n",
       "2  Drosophila melanogaster       CG12213   \n",
       "3  Drosophila melanogaster         Traf4   \n",
       "4  Drosophila melanogaster         Zfrp8   \n",
       "\n",
       "                                          annotation  \n",
       "0  LD47774p; Zinc ion binding; DNA-binding transc...  \n",
       "1  Adaptor Protein complex 1/2, beta subunit (AP-...  \n",
       "2                Uncharacterized protein, isoform C.  \n",
       "3  TNF-receptor-associated factor 4 (Traf4) encod...  \n",
       "4  Zinc finger protein RP-8 (Zfrp8) encodes a hig...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url_request = base_url + \"get_string_ids?identifiers=1%0d2%0d3%0d4%0d5\"\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request).json()\n",
    "df = pd.DataFrame(response)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ySj9iwYAr_i6"
   },
   "source": [
    "### 2.1.3. Genera una imagen del grafo de relaciones de proteínas.\n",
    "ID de proteínas: DRD2_HUMAN, DRD1_HUMAN, DRD3_HUMAN\n",
    "\n",
    "Tipo de red: funcional.\n",
    "\n",
    "La red debe mostrar TODOS los elementos independientemente de sus relaciones.\n",
    "\n",
    "¿Y si incrementamos el número de nodos asociados al conjunto de proteínas a 55?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "id": "UZHVrm28sV60"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/image/network?identifiers=DRD2_HUMAN%0dDRD1_HUMAN%0dDRD3_HUMAN&network_type=functional\n"
     ]
    }
   ],
   "source": [
    "base_url=\"https://string-db.org/api/image/\"\n",
    "url_request = base_url + \"network?identifiers=DRD2_HUMAN%0dDRD1_HUMAN%0dDRD3_HUMAN&network_type=functional\"\n",
    "\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request, stream=True)\n",
    "with open('img.png', 'wb') as out_file:\n",
    "    shutil.copyfileobj(response.raw, out_file)\n",
    "del response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "nWYWF5ynDJlC"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/image/network?identifiers=DRD2_HUMAN%0dDRD1_HUMAN%0dDRD3_HUMAN&network_type=functional&add_color_nodes=55\n"
     ]
    }
   ],
   "source": [
    "# Número de nodos considerados = 55\n",
    "url_request = base_url + \"network?identifiers=DRD2_HUMAN%0dDRD1_HUMAN%0dDRD3_HUMAN&network_type=functional&add_color_nodes=55\"\n",
    "\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request, stream=True)\n",
    "with open('img.png', 'wb') as out_file:\n",
    "    shutil.copyfileobj(response.raw, out_file)\n",
    "del response\n",
    "# .."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Uf9j-WR1sWIb"
   },
   "source": [
    "### 2.1.4. Genera una imagen del grafo de relaciones de proteínas.\n",
    "ID de proteínas: trpB\n",
    "\n",
    "El umbral de relación debe ser 70.\n",
    "\n",
    "Número de nodos: 10.\n",
    "\n",
    "Tipo de red: física.\n",
    "\n",
    "La red debe mostrar únicamente aquellos elementos que cuenten con enlaces con otras proteínas.\n",
    "\n",
    "¿Por qué no se muestran todos los nodos?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "ai4x1qbRtB8-"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/network?identifiers=trpB&network_type=physical&add_color_nodes=10&required_score=70\n"
     ]
    }
   ],
   "source": [
    "# ..\n",
    "base_url=\"https://string-db.org/api/image/\"\n",
    "url_request = base_url + \"network?identifiers=trpB&network_type=physical&add_color_nodes=10&required_score=70\"\n",
    "\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request, stream=True)\n",
    "with open('imog.png', 'wb') as out_file:\n",
    "    shutil.copyfileobj(response.raw, out_file)\n",
    "del response\n",
    "\n",
    "# ¿Por qué no se muestran todos los 10 nodos?\n",
    "# Porque no superan el umbral de relacion."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "b4bulpxXw1DG"
   },
   "source": [
    "### 2.1.5. Obtén las relaciones existentes para una proteína en particular.\n",
    "ID de la proteína: BPSL2020\n",
    "\n",
    "El umbral de relación debe ser de 40.\n",
    "\n",
    "La especie con la que se debe corresponder la proteína proporcionada es Burkholderia.\n",
    "\n",
    "Se deben mostrar 15 nodos relacionados con la proteína proporcionada.\n",
    "\n",
    "Genera la red de relaciones que generaría esta relación."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "ZHCek-8zxeDj"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/network?identifiers=BPSL2020&species=Burkholderia&add_nodes=15&required_score=40\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>272560.BPSL0019</td>\n",
       "      <td>272560.BPSS0294</td>\n",
       "      <td>BPSL0019</td>\n",
       "      <td>BPSS0294</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>272560.BPSL0019</td>\n",
       "      <td>272560.BPSS0294</td>\n",
       "      <td>BPSL0019</td>\n",
       "      <td>BPSS0294</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.157</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>272560.BPSL0019</td>\n",
       "      <td>272560.BPSS0538</td>\n",
       "      <td>BPSL0019</td>\n",
       "      <td>BPSS0538</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.211</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>272560.BPSL0019</td>\n",
       "      <td>272560.BPSS0538</td>\n",
       "      <td>BPSL0019</td>\n",
       "      <td>BPSS0538</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.211</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>272560.BPSL0019</td>\n",
       "      <td>272560.BPSS1350</td>\n",
       "      <td>BPSL0019</td>\n",
       "      <td>BPSS1350</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.211</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>272560.BPSS0624</td>\n",
       "      <td>272560.BPSS1350</td>\n",
       "      <td>macB</td>\n",
       "      <td>BPSS1350</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.181</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.081</td>\n",
       "      <td>0</td>\n",
       "      <td>0.126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>272560.BPSS0624</td>\n",
       "      <td>272560.BPSS1348</td>\n",
       "      <td>macB</td>\n",
       "      <td>BPSS1348</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.129</td>\n",
       "      <td>0</td>\n",
       "      <td>0.164</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.708</td>\n",
       "      <td>0</td>\n",
       "      <td>0.182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>272560.BPSS0624</td>\n",
       "      <td>272560.BPSS1348</td>\n",
       "      <td>macB</td>\n",
       "      <td>BPSS1348</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.808</td>\n",
       "      <td>0.129</td>\n",
       "      <td>0</td>\n",
       "      <td>0.164</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.708</td>\n",
       "      <td>0</td>\n",
       "      <td>0.182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>272560.BPSS1348</td>\n",
       "      <td>272560.BPSS1350</td>\n",
       "      <td>BPSS1348</td>\n",
       "      <td>BPSS1350</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.503</td>\n",
       "      <td>0.269</td>\n",
       "      <td>0</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>272560.BPSS1348</td>\n",
       "      <td>272560.BPSS1350</td>\n",
       "      <td>BPSS1348</td>\n",
       "      <td>BPSS1350</td>\n",
       "      <td>272560</td>\n",
       "      <td>0.503</td>\n",
       "      <td>0.269</td>\n",
       "      <td>0</td>\n",
       "      <td>0.314</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          stringId_A       stringId_B preferredName_A preferredName_B  \\\n",
       "0    272560.BPSL0019  272560.BPSS0294        BPSL0019        BPSS0294   \n",
       "1    272560.BPSL0019  272560.BPSS0294        BPSL0019        BPSS0294   \n",
       "2    272560.BPSL0019  272560.BPSS0538        BPSL0019        BPSS0538   \n",
       "3    272560.BPSL0019  272560.BPSS0538        BPSL0019        BPSS0538   \n",
       "4    272560.BPSL0019  272560.BPSS1350        BPSL0019        BPSS1350   \n",
       "..               ...              ...             ...             ...   \n",
       "191  272560.BPSS0624  272560.BPSS1350            macB        BPSS1350   \n",
       "192  272560.BPSS0624  272560.BPSS1348            macB        BPSS1348   \n",
       "193  272560.BPSS0624  272560.BPSS1348            macB        BPSS1348   \n",
       "194  272560.BPSS1348  272560.BPSS1350        BPSS1348        BPSS1350   \n",
       "195  272560.BPSS1348  272560.BPSS1350        BPSS1348        BPSS1350   \n",
       "\n",
       "    ncbiTaxonId  score  nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "0        272560  0.157   0.000       0   0.157   0.000   0.000       0   0.000  \n",
       "1        272560  0.157   0.000       0   0.157   0.000   0.000       0   0.000  \n",
       "2        272560  0.211   0.043       0   0.169   0.088   0.000       0   0.000  \n",
       "3        272560  0.211   0.043       0   0.169   0.088   0.000       0   0.000  \n",
       "4        272560  0.211   0.043       0   0.169   0.088   0.000       0   0.000  \n",
       "..          ...    ...     ...     ...     ...     ...     ...     ...     ...  \n",
       "191      272560  0.181   0.063       0   0.000   0.000   0.081       0   0.126  \n",
       "192      272560  0.808   0.129       0   0.164   0.068   0.708       0   0.182  \n",
       "193      272560  0.808   0.129       0   0.164   0.068   0.708       0   0.182  \n",
       "194      272560  0.503   0.269       0   0.314   0.088   0.000       0   0.000  \n",
       "195      272560  0.503   0.269       0   0.314   0.088   0.000       0   0.000  \n",
       "\n",
       "[196 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_url= 'https://string-db.org/api/json/'\n",
    "url_request = base_url + \"network?identifiers=BPSL2020&species=Burkholderia&add_nodes=15&required_score=40\"\n",
    "\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request).json()\n",
    "df = pd.DataFrame(response)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "C7UPg2dtFYgY"
   },
   "outputs": [],
   "source": [
    "# Generación del grafo\n",
    "base_url=\"https://string-db.org/api/image/\"\n",
    "url_request = base_url + \"network?identifiers=BPSL2020&species=Burkholderia&add_nodes=15&required_score=40\"\n",
    "response = requests.get(url_request, stream=True)\n",
    "with open('img.png', 'wb') as out_file:\n",
    "    shutil.copyfileobj(response.raw, out_file)\n",
    "del response"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "6GG7n_45xeSf"
   },
   "source": [
    "### 2.1.6. Obtén las relaciones existentes en un conjunto de proteínas.\n",
    "\n",
    "ID de las proteínas: mats, wts, hpo, trc, polo.\n",
    "\n",
    "El umbral de relación debe ser 80.\n",
    "\n",
    "La especie debe ser Drosophila melanogaster.\n",
    "\n",
    "¿Cuántas relaciones se obtienen?\n",
    "\n",
    "¿Cuál es la media de la puntuación del perfil filogenético de la red?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "id": "ekhhiuDfxm-i"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/network?identifiers=mats%0dwts%0dhpo%0dtrc%0dpolo&species=7227&add_nodes=15&required_score=80\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7227.FBpp0071822</td>\n",
       "      <td>7227.FBpp0307973</td>\n",
       "      <td>CycB</td>\n",
       "      <td>mats</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.257</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.141</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7227.FBpp0071822</td>\n",
       "      <td>7227.FBpp0307973</td>\n",
       "      <td>CycB</td>\n",
       "      <td>mats</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.257</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.141</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7227.FBpp0071822</td>\n",
       "      <td>7227.FBpp0402906</td>\n",
       "      <td>CycB</td>\n",
       "      <td>yki</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.283</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7227.FBpp0071822</td>\n",
       "      <td>7227.FBpp0402906</td>\n",
       "      <td>CycB</td>\n",
       "      <td>yki</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.283</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7227.FBpp0071822</td>\n",
       "      <td>7227.FBpp0304464</td>\n",
       "      <td>CycB</td>\n",
       "      <td>trc</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.251</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>209</th>\n",
       "      <td>7227.FBpp0306890</td>\n",
       "      <td>7227.FBpp0310068</td>\n",
       "      <td>stg</td>\n",
       "      <td>cdc14</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.708</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.047</td>\n",
       "      <td>0.527</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>7227.FBpp0307973</td>\n",
       "      <td>7227.FBpp0310068</td>\n",
       "      <td>mats</td>\n",
       "      <td>cdc14</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.770</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>7227.FBpp0307973</td>\n",
       "      <td>7227.FBpp0310068</td>\n",
       "      <td>mats</td>\n",
       "      <td>cdc14</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.770</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.154</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>7227.FBpp0307973</td>\n",
       "      <td>7227.FBpp0402906</td>\n",
       "      <td>mats</td>\n",
       "      <td>yki</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.971</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.234</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>7227.FBpp0307973</td>\n",
       "      <td>7227.FBpp0402906</td>\n",
       "      <td>mats</td>\n",
       "      <td>yki</td>\n",
       "      <td>7227</td>\n",
       "      <td>0.971</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.234</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.649</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           stringId_A        stringId_B preferredName_A preferredName_B  \\\n",
       "0    7227.FBpp0071822  7227.FBpp0307973            CycB            mats   \n",
       "1    7227.FBpp0071822  7227.FBpp0307973            CycB            mats   \n",
       "2    7227.FBpp0071822  7227.FBpp0402906            CycB             yki   \n",
       "3    7227.FBpp0071822  7227.FBpp0402906            CycB             yki   \n",
       "4    7227.FBpp0071822  7227.FBpp0304464            CycB             trc   \n",
       "..                ...               ...             ...             ...   \n",
       "209  7227.FBpp0306890  7227.FBpp0310068             stg           cdc14   \n",
       "210  7227.FBpp0307973  7227.FBpp0310068            mats           cdc14   \n",
       "211  7227.FBpp0307973  7227.FBpp0310068            mats           cdc14   \n",
       "212  7227.FBpp0307973  7227.FBpp0402906            mats             yki   \n",
       "213  7227.FBpp0307973  7227.FBpp0402906            mats             yki   \n",
       "\n",
       "    ncbiTaxonId  score  nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "0          7227  0.257       0       0     0.0   0.064   0.141     0.0   0.150  \n",
       "1          7227  0.257       0       0     0.0   0.064   0.141     0.0   0.150  \n",
       "2          7227  0.283       0       0     0.0   0.000   0.068     0.0   0.262  \n",
       "3          7227  0.283       0       0     0.0   0.000   0.068     0.0   0.262  \n",
       "4          7227  0.463       0       0     0.0   0.051   0.251     0.0   0.306  \n",
       "..          ...    ...     ...     ...     ...     ...     ...     ...     ...  \n",
       "209        7227  0.708       0       0     0.0   0.047   0.527     0.0   0.404  \n",
       "210        7227  0.770       0       0     0.0   0.000   0.154     0.0   0.739  \n",
       "211        7227  0.770       0       0     0.0   0.000   0.154     0.0   0.739  \n",
       "212        7227  0.971       0       0     0.0   0.071   0.234     0.9   0.649  \n",
       "213        7227  0.971       0       0     0.0   0.071   0.234     0.9   0.649  \n",
       "\n",
       "[214 rows x 13 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_url= 'https://string-db.org/api/json/'\n",
    "url_request = base_url + \"network?identifiers=mats%0dwts%0dhpo%0dtrc%0dpolo&species=7227&add_nodes=15&required_score=80\"\n",
    "\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request).json()\n",
    "df = pd.DataFrame(response)\n",
    "df\n",
    "# .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "id": "Yq2ETJHcHTc3"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "214"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ¿Cuántas relaciones se obtienen?\n",
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "id": "Rk-NQlZ-IYjH"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.671785046728972"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Media de Puntuación del perfil filogenético de la red.\n",
    "df['score'].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "vr8Wo08wyk0i"
   },
   "source": [
    "### 2.1.7. Devuelve la compatibilidad (homology scores) entre dos proteínas particulares.\n",
    "ID proteínas: dzip1 and osbpl1a.\n",
    "\n",
    "¿Cuál será el bitscore medio de las relaciones?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "id": "UGxxGU7byk8G"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requesting to https://string-db.org/api/json/homology?identifiers=dzip1%0dosbpl1a\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ncbiTaxonId_A</th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>ncbiTaxonId_B</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>9606.ENSP00000257312</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606.ENSP00000257312</td>\n",
       "      <td>1796.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>9606.ENSP00000320291</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606.ENSP00000320291</td>\n",
       "      <td>1983.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ncbiTaxonId_A            stringId_A  ncbiTaxonId_B            stringId_B  \\\n",
       "0           9606  9606.ENSP00000257312           9606  9606.ENSP00000257312   \n",
       "1           9606  9606.ENSP00000320291           9606  9606.ENSP00000320291   \n",
       "\n",
       "  bitscore  \n",
       "0   1796.6  \n",
       "1   1983.8  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_url= 'https://string-db.org/api/json/'\n",
    "url_request = base_url + \"homology?identifiers=dzip1%0dosbpl1a\"\n",
    "print(\"Requesting to \"+url_request)\n",
    "\n",
    "response = requests.get(url_request).json()\n",
    "df = pd.DataFrame(response)\n",
    "df\n",
    "\n",
    "# .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1890.1999999999998"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Bitscore medio\n",
    "\n",
    "df['bitscore'].astype('float64').mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dNcsu4h3DFmD"
   },
   "source": [
    "## 2.2. Consultas mediante stringdb."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "53JjwSJyh-3y"
   },
   "source": [
    "La librería de Python stringdb proporciona un conjunto de operaciones que interactúan con la API Rest de STRING para ofrecer resultados. Para ello, se puede hacer uso de cinco métodos:\n",
    "- get_enrichement: devuelve las relaciones funcionales del pathways del conjunto de proteínas proporcionadas como parámetro.\n",
    "- get_interaction_partners: proporciona las interacciones identificadas para las proteínas indicadas.\n",
    "- get_ppi_enrichment: calcula el valor de enriquecimiento de los valores indicados, es decir, el denominado Protein-Protein Interaction (PPI) enrichment (p-value), referencia que mide el nivel de interacción entre las entidades.\n",
    "- get_network: devuelve la red de relaciones generadas entre las proteínas indicadas.\n",
    "- get_functional_annotation: determina todos los pathways para la lista de proteínas indicadas.\n",
    "- get_string_ids: muestra las características correspondientes a las proteínas proporcionadas como parámetro.\n",
    "Adicionalmente, se proporciona un método que permite invocar directamente peticiones mediante una URL a los _endpoints_ de la API.\n",
    "\n",
    "Es interesante considerar que los métodos proporcionan un DataFrame como resultado. De esta manera, será posible realizar operaciones de ordenado y clasificación, así como recorrido, de la información.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rYFCLsWO3eA-"
   },
   "source": [
    "### 2.2.1. Configuración de la librería."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MY3ro1nj39ix"
   },
   "source": [
    "#### 2.2.1.1. Instala la librería stringdb."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "id": "Ax18sNergIqF"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: stringdb in c:\\users\\carlos\\anaconda3\\lib\\site-packages (0.1.5)\n",
      "Requirement already satisfied: pandas in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from stringdb) (1.5.3)\n",
      "Requirement already satisfied: requests in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from stringdb) (2.29.0)\n",
      "Requirement already satisfied: numpy>=1.20.3 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from pandas->stringdb) (1.23.5)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from pandas->stringdb) (2022.7)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from pandas->stringdb) (2.8.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from requests->stringdb) (2.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from requests->stringdb) (2023.5.7)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from requests->stringdb) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from requests->stringdb) (1.26.15)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\carlos\\anaconda3\\lib\\site-packages (from python-dateutil>=2.8.1->pandas->stringdb) (1.16.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install stringdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "oZ6RGYxr3s2m"
   },
   "source": [
    "#### 2.2.1.2. Importa la librería stringdb.\n",
    "Tras este paso, ya podrás hacer uso de las funciones que proporciona la librería."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "vPGsBIlL32as"
   },
   "source": [
    "### 2.2.3. Ejercicios."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "gq3uqPn34FDJ"
   },
   "source": [
    "#### 2.2.3.1. Obtén las características completas de un conjunto de genes.\n",
    "¿A qué especie pertenecen?\n",
    "Colección de genes: 'TP53', 'BRCA1', 'FANCD1', 'FANCL'.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "id": "43VlrinAgDIk"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>queryItem</th>\n",
       "      <th>queryIndex</th>\n",
       "      <th>stringId</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>taxonName</th>\n",
       "      <th>preferredName</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>0</td>\n",
       "      <td>9606.ENSP00000269305</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Cellular tumor antigen p53; Acts as a tumor su...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>1</td>\n",
       "      <td>9606.ENSP00000418960</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Breast cancer type 1 susceptibility protein; E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FANCD1</td>\n",
       "      <td>2</td>\n",
       "      <td>9606.ENSP00000369497</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>Breast cancer type 2 susceptibility protein; I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FANCL</td>\n",
       "      <td>3</td>\n",
       "      <td>9606.ENSP00000385021</td>\n",
       "      <td>9606</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FANCL</td>\n",
       "      <td>E3 ubiquitin-protein ligase FANCL; Ubiquitin l...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  queryItem  queryIndex              stringId  ncbiTaxonId     taxonName  \\\n",
       "0      TP53           0  9606.ENSP00000269305         9606  Homo sapiens   \n",
       "1     BRCA1           1  9606.ENSP00000418960         9606  Homo sapiens   \n",
       "2    FANCD1           2  9606.ENSP00000369497         9606  Homo sapiens   \n",
       "3     FANCL           3  9606.ENSP00000385021         9606  Homo sapiens   \n",
       "\n",
       "  preferredName                                         annotation  \n",
       "0          TP53  Cellular tumor antigen p53; Acts as a tumor su...  \n",
       "1         BRCA1  Breast cancer type 1 susceptibility protein; E...  \n",
       "2         BRCA2  Breast cancer type 2 susceptibility protein; I...  \n",
       "3         FANCL  E3 ubiquitin-protein ligase FANCL; Ubiquitin l...  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Características de los genes.\n",
    "genes =['TP53', 'BRCA1', 'FANCD1', 'FANCL']\n",
    "string_ids=stringdb.get_string_ids(genes)\n",
    "string_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "id": "6OLe6_9-6Lh1"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    Homo sapiens\n",
       "1    Homo sapiens\n",
       "2    Homo sapiens\n",
       "3    Homo sapiens\n",
       "Name: taxonName, dtype: object"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Especie a la que pertenecen:\n",
    "\n",
    "string_ids['taxonName']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JiWoY4G94fda"
   },
   "source": [
    "#### 2.2.3.2. Obtén una lista de las relaciones funcionales para una lista de proteínas.\n",
    "Colección de proteínas: 'WDR70', 'PLRG1', 'DDB1', 'CUL4A'.\n",
    "\n",
    "¿Hay alguna enfermedad entre las relaciones?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "id": "5hYrx5M9gH-7"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>term</th>\n",
       "      <th>number_of_genes</th>\n",
       "      <th>number_of_genes_in_background</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>inputGenes</th>\n",
       "      <th>preferredNames</th>\n",
       "      <th>p_value</th>\n",
       "      <th>fdr</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Component</td>\n",
       "      <td>GO:0031464</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>3.240000e-06</td>\n",
       "      <td>0.0066</td>\n",
       "      <td>Cul4A-RING E3 ubiquitin ligase complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COMPARTMENTS</td>\n",
       "      <td>GOCC:0031464</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>4.200000e-06</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>Cul4A-RING E3 ubiquitin ligase complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>hsa03420</td>\n",
       "      <td>2</td>\n",
       "      <td>46</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>3.480000e-05</td>\n",
       "      <td>0.0117</td>\n",
       "      <td>Nucleotide excision repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMID</td>\n",
       "      <td>PMID:30791110</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>9606</td>\n",
       "      <td>WDR70,DDB1,CUL4A</td>\n",
       "      <td>WDR70,DDB1,CUL4A</td>\n",
       "      <td>4.170000e-09</td>\n",
       "      <td>0.0162</td>\n",
       "      <td>(2019) The Antiresection Activity of the X Pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PMID</td>\n",
       "      <td>PMID:25461807</td>\n",
       "      <td>3</td>\n",
       "      <td>26</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A,PLRG1</td>\n",
       "      <td>DDB1,CUL4A,PLRG1</td>\n",
       "      <td>1.150000e-08</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>(2015) BCAS2 promotes prostate cancer cells pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>WikiPathways</td>\n",
       "      <td>WP4753</td>\n",
       "      <td>2</td>\n",
       "      <td>44</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>3.190000e-05</td>\n",
       "      <td>0.0249</td>\n",
       "      <td>Nucleotide excision repair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PMID</td>\n",
       "      <td>PMID:33627782</td>\n",
       "      <td>3</td>\n",
       "      <td>33</td>\n",
       "      <td>9606</td>\n",
       "      <td>WDR70,DDB1,CUL4A</td>\n",
       "      <td>WDR70,DDB1,CUL4A</td>\n",
       "      <td>2.240000e-08</td>\n",
       "      <td>0.0289</td>\n",
       "      <td>(2021) CRL4A(DTL) degrades DNA-PKcs to modulat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-6782210</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>6.600000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>Gap-filling DNA repair synthesis and ligation ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-6782135</td>\n",
       "      <td>2</td>\n",
       "      <td>65</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>6.800000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>Dual incision in TC-NER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-6781823</td>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>4.570000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>Formation of TC-NER Pre-Incision Complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-5696400</td>\n",
       "      <td>2</td>\n",
       "      <td>41</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>2.780000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>Dual Incision in GG-NER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-110314</td>\n",
       "      <td>2</td>\n",
       "      <td>30</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>1.530000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>Recognition of DNA damage by PCNA-containing r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-5696394</td>\n",
       "      <td>2</td>\n",
       "      <td>38</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>2.410000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>DNA Damage Recognition in GG-NER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RCTM</td>\n",
       "      <td>HSA-5696395</td>\n",
       "      <td>2</td>\n",
       "      <td>42</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>2.920000e-05</td>\n",
       "      <td>0.0349</td>\n",
       "      <td>Formation of Incision Complex in GG-NER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>WikiPathways</td>\n",
       "      <td>WP5114</td>\n",
       "      <td>2</td>\n",
       "      <td>75</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>9.000000e-05</td>\n",
       "      <td>0.0351</td>\n",
       "      <td>Nucleotide excision repair in xeroderma pigmen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NetworkNeighborAL</td>\n",
       "      <td>CL:4414</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>7.810000e-06</td>\n",
       "      <td>0.0357</td>\n",
       "      <td>Cul4-RING E3 ubiquitin ligase complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PMID</td>\n",
       "      <td>PMID:20658650</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A,PLRG1</td>\n",
       "      <td>DDB1,CUL4A,PLRG1</td>\n",
       "      <td>4.450000e-08</td>\n",
       "      <td>0.0431</td>\n",
       "      <td>(2010) Initiation of DNA interstrand cross-lin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KEGG</td>\n",
       "      <td>hsa04120</td>\n",
       "      <td>2</td>\n",
       "      <td>134</td>\n",
       "      <td>9606</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>DDB1,CUL4A</td>\n",
       "      <td>2.800000e-04</td>\n",
       "      <td>0.0472</td>\n",
       "      <td>Ubiquitin mediated proteolysis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             category           term  number_of_genes  \\\n",
       "1           Component     GO:0031464                2   \n",
       "0        COMPARTMENTS   GOCC:0031464                2   \n",
       "2                KEGG       hsa03420                2   \n",
       "4                PMID  PMID:30791110                3   \n",
       "5                PMID  PMID:25461807                3   \n",
       "15       WikiPathways         WP4753                2   \n",
       "6                PMID  PMID:33627782                3   \n",
       "14               RCTM    HSA-6782210                2   \n",
       "13               RCTM    HSA-6782135                2   \n",
       "12               RCTM    HSA-6781823                2   \n",
       "11               RCTM    HSA-5696400                2   \n",
       "8                RCTM     HSA-110314                2   \n",
       "9                RCTM    HSA-5696394                2   \n",
       "10               RCTM    HSA-5696395                2   \n",
       "16       WikiPathways         WP5114                2   \n",
       "17  NetworkNeighborAL        CL:4414                2   \n",
       "7                PMID  PMID:20658650                3   \n",
       "3                KEGG       hsa04120                2   \n",
       "\n",
       "    number_of_genes_in_background  ncbiTaxonId        inputGenes  \\\n",
       "1                              13         9606        DDB1,CUL4A   \n",
       "0                              15         9606        DDB1,CUL4A   \n",
       "2                              46         9606        DDB1,CUL4A   \n",
       "4                              18         9606  WDR70,DDB1,CUL4A   \n",
       "5                              26         9606  DDB1,CUL4A,PLRG1   \n",
       "15                             44         9606        DDB1,CUL4A   \n",
       "6                              33         9606  WDR70,DDB1,CUL4A   \n",
       "14                             64         9606        DDB1,CUL4A   \n",
       "13                             65         9606        DDB1,CUL4A   \n",
       "12                             53         9606        DDB1,CUL4A   \n",
       "11                             41         9606        DDB1,CUL4A   \n",
       "8                              30         9606        DDB1,CUL4A   \n",
       "9                              38         9606        DDB1,CUL4A   \n",
       "10                             42         9606        DDB1,CUL4A   \n",
       "16                             75         9606        DDB1,CUL4A   \n",
       "17                             21         9606        DDB1,CUL4A   \n",
       "7                              42         9606  DDB1,CUL4A,PLRG1   \n",
       "3                             134         9606        DDB1,CUL4A   \n",
       "\n",
       "      preferredNames       p_value     fdr  \\\n",
       "1         DDB1,CUL4A  3.240000e-06  0.0066   \n",
       "0         DDB1,CUL4A  4.200000e-06  0.0097   \n",
       "2         DDB1,CUL4A  3.480000e-05  0.0117   \n",
       "4   WDR70,DDB1,CUL4A  4.170000e-09  0.0162   \n",
       "5   DDB1,CUL4A,PLRG1  1.150000e-08  0.0222   \n",
       "15        DDB1,CUL4A  3.190000e-05  0.0249   \n",
       "6   WDR70,DDB1,CUL4A  2.240000e-08  0.0289   \n",
       "14        DDB1,CUL4A  6.600000e-05  0.0349   \n",
       "13        DDB1,CUL4A  6.800000e-05  0.0349   \n",
       "12        DDB1,CUL4A  4.570000e-05  0.0349   \n",
       "11        DDB1,CUL4A  2.780000e-05  0.0349   \n",
       "8         DDB1,CUL4A  1.530000e-05  0.0349   \n",
       "9         DDB1,CUL4A  2.410000e-05  0.0349   \n",
       "10        DDB1,CUL4A  2.920000e-05  0.0349   \n",
       "16        DDB1,CUL4A  9.000000e-05  0.0351   \n",
       "17        DDB1,CUL4A  7.810000e-06  0.0357   \n",
       "7   DDB1,CUL4A,PLRG1  4.450000e-08  0.0431   \n",
       "3         DDB1,CUL4A  2.800000e-04  0.0472   \n",
       "\n",
       "                                          description  \n",
       "1              Cul4A-RING E3 ubiquitin ligase complex  \n",
       "0              Cul4A-RING E3 ubiquitin ligase complex  \n",
       "2                          Nucleotide excision repair  \n",
       "4   (2019) The Antiresection Activity of the X Pro...  \n",
       "5   (2015) BCAS2 promotes prostate cancer cells pr...  \n",
       "15                         Nucleotide excision repair  \n",
       "6   (2021) CRL4A(DTL) degrades DNA-PKcs to modulat...  \n",
       "14  Gap-filling DNA repair synthesis and ligation ...  \n",
       "13                            Dual incision in TC-NER  \n",
       "12           Formation of TC-NER Pre-Incision Complex  \n",
       "11                            Dual Incision in GG-NER  \n",
       "8   Recognition of DNA damage by PCNA-containing r...  \n",
       "9                    DNA Damage Recognition in GG-NER  \n",
       "10            Formation of Incision Complex in GG-NER  \n",
       "16  Nucleotide excision repair in xeroderma pigmen...  \n",
       "17              Cul4-RING E3 ubiquitin ligase complex  \n",
       "7   (2010) Initiation of DNA interstrand cross-lin...  \n",
       "3                      Ubiquitin mediated proteolysis  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ['WDR70', 'PLRG1', 'DDB1', 'CUL4A']\n",
    "\n",
    "enrichment_df = stringdb.get_enrichment(genes)\n",
    "enrichment_df.sort_values('fdr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "id": "aXG1ss346F6Z"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Número de enfermedades ('DISEASES') en el conjunto de salida.\n",
    "diseases = 0\n",
    "for mol in enrichment_df['category']:\n",
    "    if(mol == 'DISEASES'):\n",
    "        diseases = diseases + 1\n",
    "\n",
    "diseases\n",
    "# .."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "kb9JZv7i6hHb"
   },
   "source": [
    "#### 2.2.3.3. Genera el conjunto de relaciones asociadas a un conjunto de genes.\n",
    "Colección de genes: 'CUL4B', 'RBX1', 'SNW1', 'CDC40'.\n",
    "\n",
    "El umbral de compatibilidades (required_score) debe ser 400.\n",
    "\n",
    "La especie de las proteínas (species) involucradas debe ser humana (9606).\n",
    "\n",
    "¿Cuántas relaciones se han calculado?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "id": "vB-WtXNZ7KDb"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000216225</td>\n",
       "      <td>9606.ENSP00000451129</td>\n",
       "      <td>RBX1</td>\n",
       "      <td>SNW1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.585</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000216225</td>\n",
       "      <td>9606.ENSP00000451129</td>\n",
       "      <td>RBX1</td>\n",
       "      <td>SNW1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.585</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.064</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000216225</td>\n",
       "      <td>9606.ENSP00000384109</td>\n",
       "      <td>RBX1</td>\n",
       "      <td>CUL4B</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000216225</td>\n",
       "      <td>9606.ENSP00000384109</td>\n",
       "      <td>RBX1</td>\n",
       "      <td>CUL4B</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.993</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000357928</td>\n",
       "      <td>9606.ENSP00000451129</td>\n",
       "      <td>CDC40</td>\n",
       "      <td>SNW1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9606.ENSP00000357928</td>\n",
       "      <td>9606.ENSP00000451129</td>\n",
       "      <td>CDC40</td>\n",
       "      <td>SNW1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.169</td>\n",
       "      <td>0.997</td>\n",
       "      <td>0.9</td>\n",
       "      <td>0.763</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             stringId_A            stringId_B preferredName_A preferredName_B  \\\n",
       "0  9606.ENSP00000216225  9606.ENSP00000451129            RBX1            SNW1   \n",
       "1  9606.ENSP00000216225  9606.ENSP00000451129            RBX1            SNW1   \n",
       "2  9606.ENSP00000216225  9606.ENSP00000384109            RBX1           CUL4B   \n",
       "3  9606.ENSP00000216225  9606.ENSP00000384109            RBX1           CUL4B   \n",
       "4  9606.ENSP00000357928  9606.ENSP00000451129           CDC40            SNW1   \n",
       "5  9606.ENSP00000357928  9606.ENSP00000451129           CDC40            SNW1   \n",
       "\n",
       "   ncbiTaxonId  score  nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "0         9606  0.585       0       0       0   0.064   0.000     0.5   0.186  \n",
       "1         9606  0.585       0       0       0   0.064   0.000     0.5   0.186  \n",
       "2         9606  0.999       0       0       0   0.074   0.993     0.9   0.996  \n",
       "3         9606  0.999       0       0       0   0.074   0.993     0.9   0.996  \n",
       "4         9606  0.999       0       0       0   0.169   0.997     0.9   0.763  \n",
       "5         9606  0.999       0       0       0   0.169   0.997     0.9   0.763  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ['CUL4B', 'RBX1', 'SNW1', 'CDC40']\n",
    "\n",
    "enrichment_df = stringdb.get_network(genes, required_score=400, species=9606)\n",
    "enrichment_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "id": "IHi86Dn27uOc"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Número de relaciones obtenidas.\n",
    "len(enrichment_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "otNPLHwW7eT7"
   },
   "source": [
    "#### 2.2.3.4. Determina las interacciones entre las proteínas proporcionadas, así como el resto de moléculas de STRING.\n",
    "Colección de genes: 'C2orf69', 'TYW5', 'CFAP300', 'ANKRD13C', 'MOSMO', 'ENSANIP00000001692'.\n",
    "\n",
    "Determina cuál es la relación con el valor máximo de Text mining score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "c4pqyUTl7gNI"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000312770</td>\n",
       "      <td>9606.ENSP00000359982</td>\n",
       "      <td>C2orf69</td>\n",
       "      <td>ANKRD13C</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.144</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000312770</td>\n",
       "      <td>9606.ENSP00000359982</td>\n",
       "      <td>C2orf69</td>\n",
       "      <td>ANKRD13C</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.144</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000312770</td>\n",
       "      <td>9606.ENSP00000346627</td>\n",
       "      <td>C2orf69</td>\n",
       "      <td>TYW5</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000312770</td>\n",
       "      <td>9606.ENSP00000346627</td>\n",
       "      <td>C2orf69</td>\n",
       "      <td>TYW5</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.080</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.582</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             stringId_A            stringId_B preferredName_A preferredName_B  \\\n",
       "0  9606.ENSP00000312770  9606.ENSP00000359982         C2orf69        ANKRD13C   \n",
       "1  9606.ENSP00000312770  9606.ENSP00000359982         C2orf69        ANKRD13C   \n",
       "2  9606.ENSP00000312770  9606.ENSP00000346627         C2orf69            TYW5   \n",
       "3  9606.ENSP00000312770  9606.ENSP00000346627         C2orf69            TYW5   \n",
       "\n",
       "   ncbiTaxonId  score  nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "0         9606  0.490       0       0       0   0.144       0       0   0.428  \n",
       "1         9606  0.490       0       0       0   0.144       0       0   0.428  \n",
       "2         9606  0.598       0       0       0   0.080       0       0   0.582  \n",
       "3         9606  0.598       0       0       0   0.080       0       0   0.582  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ['C2orf69', 'TYW5', 'CFAP300', 'ANKRD13C', 'MOSMO', 'ENSANIP00000001692']\n",
    "\n",
    "enrichment_df = stringdb.get_network(genes)\n",
    "enrichment_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "Kse35nRHLCuL"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000312770</td>\n",
       "      <td>9606.ENSP00000346627</td>\n",
       "      <td>C2orf69</td>\n",
       "      <td>TYW5</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000312770</td>\n",
       "      <td>9606.ENSP00000346627</td>\n",
       "      <td>C2orf69</td>\n",
       "      <td>TYW5</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.582</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             stringId_A            stringId_B preferredName_A preferredName_B  \\\n",
       "2  9606.ENSP00000312770  9606.ENSP00000346627         C2orf69            TYW5   \n",
       "3  9606.ENSP00000312770  9606.ENSP00000346627         C2orf69            TYW5   \n",
       "\n",
       "   ncbiTaxonId  score  nscore  fscore  pscore  ascore  escore  dscore  tscore  \n",
       "2         9606  0.598       0       0       0    0.08       0       0   0.582  \n",
       "3         9606  0.598       0       0       0    0.08       0       0   0.582  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fila con el valor de text mining score más alto.\n",
    "enrichment_df.loc[enrichment_df['tscore']==enrichment_df['tscore'].max()]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "5auf-TZc8Rbx"
   },
   "source": [
    "#### 2.2.3.5. Calcula el valor de interacciones en un conjunto de proteínas.\n",
    "Colección de genes: 'TMEM69', 'PRDX1', 'LRRC7', 'ENSANIP00000009605', 'NEK5'.\n",
    "\n",
    "¿Cuál es el grado de relaciones medio por nodo?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "id": "sO7Z6DOrg3YM"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>stringId_A</th>\n",
       "      <th>stringId_B</th>\n",
       "      <th>preferredName_A</th>\n",
       "      <th>preferredName_B</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>score</th>\n",
       "      <th>nscore</th>\n",
       "      <th>fscore</th>\n",
       "      <th>pscore</th>\n",
       "      <th>ascore</th>\n",
       "      <th>escore</th>\n",
       "      <th>dscore</th>\n",
       "      <th>tscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606.ENSP00000262746</td>\n",
       "      <td>9606.ENSP00000371388</td>\n",
       "      <td>PRDX1</td>\n",
       "      <td>SRXN1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.999</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.131</td>\n",
       "      <td>0.957</td>\n",
       "      <td>0.75</td>\n",
       "      <td>0.942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606.ENSP00000262746</td>\n",
       "      <td>9606.ENSP00000363641</td>\n",
       "      <td>PRDX1</td>\n",
       "      <td>TXN</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.289</td>\n",
       "      <td>0.720</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606.ENSP00000262746</td>\n",
       "      <td>9606.ENSP00000265462</td>\n",
       "      <td>PRDX1</td>\n",
       "      <td>PRDX5</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.981</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.138</td>\n",
       "      <td>0.656</td>\n",
       "      <td>0.40</td>\n",
       "      <td>0.908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606.ENSP00000262746</td>\n",
       "      <td>9606.ENSP00000216185</td>\n",
       "      <td>PRDX1</td>\n",
       "      <td>TXN2</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.928</td>\n",
       "      <td>0.048</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.139</td>\n",
       "      <td>0.174</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606.ENSP00000262746</td>\n",
       "      <td>9606.ENSP00000361423</td>\n",
       "      <td>PRDX1</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.927</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.074</td>\n",
       "      <td>0.251</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>9606.ENSP00000361095</td>\n",
       "      <td>9606.ENSP00000417138</td>\n",
       "      <td>TMEM69</td>\n",
       "      <td>PLTP</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.425</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>9606.ENSP00000361095</td>\n",
       "      <td>9606.ENSP00000321810</td>\n",
       "      <td>TMEM69</td>\n",
       "      <td>TRIT1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.415</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.249</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>9606.ENSP00000361095</td>\n",
       "      <td>9606.ENSP00000318029</td>\n",
       "      <td>TMEM69</td>\n",
       "      <td>ODF3L2</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.415</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>358</th>\n",
       "      <td>9606.ENSP00000361095</td>\n",
       "      <td>9606.ENSP00000430025</td>\n",
       "      <td>TMEM69</td>\n",
       "      <td>LYRM2</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.415</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.063</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>9606.ENSP00000361095</td>\n",
       "      <td>9606.ENSP00000425631</td>\n",
       "      <td>TMEM69</td>\n",
       "      <td>LIMCH1</td>\n",
       "      <td>9606</td>\n",
       "      <td>0.410</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.410</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>360 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               stringId_A            stringId_B preferredName_A  \\\n",
       "0    9606.ENSP00000262746  9606.ENSP00000371388           PRDX1   \n",
       "1    9606.ENSP00000262746  9606.ENSP00000363641           PRDX1   \n",
       "2    9606.ENSP00000262746  9606.ENSP00000265462           PRDX1   \n",
       "3    9606.ENSP00000262746  9606.ENSP00000216185           PRDX1   \n",
       "4    9606.ENSP00000262746  9606.ENSP00000361423           PRDX1   \n",
       "..                    ...                   ...             ...   \n",
       "355  9606.ENSP00000361095  9606.ENSP00000417138          TMEM69   \n",
       "356  9606.ENSP00000361095  9606.ENSP00000321810          TMEM69   \n",
       "357  9606.ENSP00000361095  9606.ENSP00000318029          TMEM69   \n",
       "358  9606.ENSP00000361095  9606.ENSP00000430025          TMEM69   \n",
       "359  9606.ENSP00000361095  9606.ENSP00000425631          TMEM69   \n",
       "\n",
       "    preferredName_B  ncbiTaxonId  score  nscore  fscore  pscore  ascore  \\\n",
       "0             SRXN1         9606  0.999   0.000       0     0.0   0.131   \n",
       "1               TXN         9606  0.990   0.048       0     0.0   0.289   \n",
       "2             PRDX5         9606  0.981   0.000       0     0.0   0.138   \n",
       "3              TXN2         9606  0.928   0.048       0     0.0   0.139   \n",
       "4              ABL1         9606  0.927   0.000       0     0.0   0.074   \n",
       "..              ...          ...    ...     ...     ...     ...     ...   \n",
       "355            PLTP         9606  0.425   0.000       0     0.0   0.000   \n",
       "356           TRIT1         9606  0.415   0.000       0     0.0   0.249   \n",
       "357          ODF3L2         9606  0.415   0.000       0     0.0   0.000   \n",
       "358           LYRM2         9606  0.415   0.000       0     0.0   0.063   \n",
       "359          LIMCH1         9606  0.410   0.000       0     0.0   0.000   \n",
       "\n",
       "     escore  dscore  tscore  \n",
       "0     0.957    0.75   0.942  \n",
       "1     0.720    0.50   0.915  \n",
       "2     0.656    0.40   0.908  \n",
       "3     0.174    0.00   0.906  \n",
       "4     0.251    0.00   0.904  \n",
       "..      ...     ...     ...  \n",
       "355   0.000    0.00   0.425  \n",
       "356   0.000    0.00   0.253  \n",
       "357   0.000    0.00   0.415  \n",
       "358   0.000    0.00   0.401  \n",
       "359   0.000    0.00   0.410  \n",
       "\n",
       "[360 rows x 13 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ['TMEM69', 'PRDX1', 'LRRC7', 'ENSANIP00000009605', 'NEK5']\n",
    "interaction_partners_df = stringdb.get_interaction_partners(genes)\n",
    "interaction_partners_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number_of_nodes</th>\n",
       "      <th>number_of_edges</th>\n",
       "      <th>average_node_degree</th>\n",
       "      <th>local_clustering_coefficient</th>\n",
       "      <th>expected_number_of_edges</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>51</td>\n",
       "      <td>9.27</td>\n",
       "      <td>0.926</td>\n",
       "      <td>19</td>\n",
       "      <td>6.210000e-10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   number_of_nodes  number_of_edges  average_node_degree  \\\n",
       "0               11               51                 9.27   \n",
       "\n",
       "   local_clustering_coefficient  expected_number_of_edges       p_value  \n",
       "0                         0.926                        19  6.210000e-10  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interaction_partners_df = stringdb.get_ppi_enrichment(genes)\n",
    "interaction_partners_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Average node degree: 9.27"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "44yYLFBM87sX"
   },
   "source": [
    "#### 2.2.3.6. Determina los pathways de un gen.\n",
    "Colección de genes: 'BRCA1'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "id": "VACXb4b2g_IR"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>term</th>\n",
       "      <th>number_of_genes</th>\n",
       "      <th>ratio_in_set</th>\n",
       "      <th>ncbiTaxonId</th>\n",
       "      <th>inputGenes</th>\n",
       "      <th>preferredNames</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COMPARTMENTS</td>\n",
       "      <td>GOCC:0000151</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Ubiquitin ligase complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COMPARTMENTS</td>\n",
       "      <td>GOCC:0000152</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Nuclear ubiquitin ligase complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COMPARTMENTS</td>\n",
       "      <td>GOCC:0000228</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Nuclear chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COMPARTMENTS</td>\n",
       "      <td>GOCC:0000793</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Condensed chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COMPARTMENTS</td>\n",
       "      <td>GOCC:0000794</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Condensed nuclear chromosome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1121</th>\n",
       "      <td>NetworkNeighborAL</td>\n",
       "      <td>CL:6149</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Homologous DNA Pairing and Strand Exchange, an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1122</th>\n",
       "      <td>NetworkNeighborAL</td>\n",
       "      <td>CL:6150</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Resolution of D-loop Structures through Hollid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1123</th>\n",
       "      <td>NetworkNeighborAL</td>\n",
       "      <td>CL:6195</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1-C complex, and NIPBL role in DNA damage ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124</th>\n",
       "      <td>NetworkNeighborAL</td>\n",
       "      <td>CL:6207</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1-C complex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1125</th>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF12820</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>9606</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>Serine-rich domain associated with BRCT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1126 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               category          term  number_of_genes  ratio_in_set  \\\n",
       "0          COMPARTMENTS  GOCC:0000151                1           1.0   \n",
       "1          COMPARTMENTS  GOCC:0000152                1           1.0   \n",
       "2          COMPARTMENTS  GOCC:0000228                1           1.0   \n",
       "3          COMPARTMENTS  GOCC:0000793                1           1.0   \n",
       "4          COMPARTMENTS  GOCC:0000794                1           1.0   \n",
       "...                 ...           ...              ...           ...   \n",
       "1121  NetworkNeighborAL       CL:6149                1           1.0   \n",
       "1122  NetworkNeighborAL       CL:6150                1           1.0   \n",
       "1123  NetworkNeighborAL       CL:6195                1           1.0   \n",
       "1124  NetworkNeighborAL       CL:6207                1           1.0   \n",
       "1125               Pfam       PF12820                1           1.0   \n",
       "\n",
       "      ncbiTaxonId inputGenes preferredNames  \\\n",
       "0            9606      BRCA1          BRCA1   \n",
       "1            9606      BRCA1          BRCA1   \n",
       "2            9606      BRCA1          BRCA1   \n",
       "3            9606      BRCA1          BRCA1   \n",
       "4            9606      BRCA1          BRCA1   \n",
       "...           ...        ...            ...   \n",
       "1121         9606      BRCA1          BRCA1   \n",
       "1122         9606      BRCA1          BRCA1   \n",
       "1123         9606      BRCA1          BRCA1   \n",
       "1124         9606      BRCA1          BRCA1   \n",
       "1125         9606      BRCA1          BRCA1   \n",
       "\n",
       "                                            description  \n",
       "0                              Ubiquitin ligase complex  \n",
       "1                      Nuclear ubiquitin ligase complex  \n",
       "2                                    Nuclear chromosome  \n",
       "3                                  Condensed chromosome  \n",
       "4                          Condensed nuclear chromosome  \n",
       "...                                                 ...  \n",
       "1121  Homologous DNA Pairing and Strand Exchange, an...  \n",
       "1122  Resolution of D-loop Structures through Hollid...  \n",
       "1123  BRCA1-C complex, and NIPBL role in DNA damage ...  \n",
       "1124                                    BRCA1-C complex  \n",
       "1125            Serine-rich domain associated with BRCT  \n",
       "\n",
       "[1126 rows x 8 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = ['BRCA1']\n",
    "interaction_partners_df = stringdb.get_functional_annotation(genes)\n",
    "interaction_partners_df"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
